src/m3ninx/index/regexp.go

```diff
@@ -0,0 +1,323 @@
+			}
+		}
+		// strip right-most '$'
+		if l := len(parsed.Sub); rightmost && l > 0 {
+			newRe, c, err := ensureRegexpUnanchoredHelper(parsed.Sub[l-1], leftmost && l == 1, rightmost)
+			if err != nil {
+				return nil, false, err
+			}
+			if c {
+				parsed.Sub[l-1] = newRe
+				changed = true
+			}
+		}
+	case syntax.OpAlternate:
+		// strip left-most '^' and right-most '$' in each sub-expression
+		for idx := range parsed.Sub {
+			newRe, c, err := ensureRegexpUnanchoredHelper(parsed.Sub[idx], leftmost, rightmost)
+			if err != nil {
+				return nil, false, err
+			}
+			parsed.Sub[idx] = newRe
+			changed = changed || c
+		}
+	case syntax.OpQuest:
+		if len(parsed.Sub) > 0 {
+			newRe, c, err := ensureRegexpUnanchoredHelper(parsed.Sub[0], leftmost, rightmost)
+			if err != nil {
+				return nil, false, err
+			}
+			if c {
+				parsed.Sub[0] = newRe
+				changed = true
+			}
+		}
+	case syntax.OpStar:
+		if len(parsed.Sub) > 0 {
+			original := deepCopy(parsed)
+			newRe, c, err := ensureRegexpUnanchoredHelper(parsed.Sub[0], leftmost, rightmost)
+			if err != nil {
+				return nil, false, err
+			}
+			if !c {
+				return parsed, false, nil
+			}
+			newRegexp = &syntax.Regexp{
+				Op:    syntax.OpConcat,
+				Flags: parsed.Flags,
+				Sub: []*syntax.Regexp{
+					&syntax.Regexp{
+						Op:    syntax.OpQuest,
+						Flags: parsed.Flags,
+						Sub: []*syntax.Regexp{
+							newRe,
+						},
+					},
+					original,
+				},
+			}
+			changed = true
+		}
+	case syntax.OpPlus:
+		if len(parsed.Sub) > 0 {
+			original := deepCopy(parsed)
+			newRe, c, err := ensureRegexpUnanchoredHelper(parsed.Sub[0], leftmost, rightmost)
+			if err != nil {
+				return nil, false, err
+			}
+			if !c {
+				return parsed, false, nil
+			}
+			return &syntax.Regexp{
+				Op:    syntax.OpConcat,
+				Flags: parsed.Flags,
+				Sub: []*syntax.Regexp{
+					newRe,
+					&syntax.Regexp{
+						Op:    syntax.OpStar,
+						Flags: parsed.Flags,
+						Sub: []*syntax.Regexp{
+							original.Sub[0],
+						},
+					},
+				},
+			}, true, nil
+		}
+	case syntax.OpRepeat:
+		if len(parsed.Sub) > 0 && parsed.Min > 0 && parsed.Max > 0 {
+			original := deepCopy(parsed)
+			newRe, c, err := ensureRegexpUnanchoredHelper(parsed.Sub[0], leftmost, rightmost)
+			if err != nil {
+				return nil, false, err
+			}
+			if !c {
+				return parsed, false, nil
+			}
+			original.Min--
+			original.Max--
+			newRegexp = &syntax.Regexp{
+				Op:    syntax.OpConcat,
+				Flags: parsed.Flags,
+				Sub: []*syntax.Regexp{
+					newRe,
+					original,
+				},
+			}
+			changed = true
+		}
+	}
+	return newRegexp, changed, nil
+}
+
+func deepCopy(ast *syntax.Regexp) *syntax.Regexp {
+	if ast == nil {
+		return nil
+	}
+	copied := *ast
+	copied.Sub = make([]*syntax.Regexp, 0, len(ast.Sub))
+	for _, r := range ast.Sub {
+		copied.Sub = append(copied.Sub, deepCopy(r))
+	}
+	if len(copied.Sub0) != 0 && copied.Sub0[0] != nil {
+		copied.Sub0[0] = deepCopy(copied.Sub0[0])
+	}
+	return &copied
```

Comment:
Do we need copy `ast.Rune` slice?  It is apparently a `Rune []rune` meaning it could be a heap allocated slice depending on if it outgrows `ast.Rune0` array.

Reply:
only reason we don't is cause we never modify the runes, i'll add a comment to clarify

Reply:
done